### PR TITLE
Introduce apt repository config and deb resources

### PR DESCRIPTION
**Pull request by TeodorPt**
This PR introduces the possibility to optionally deploy from a custom apt repository, a custom TImescaleDB version, or from custom deb resource. By default, it will pull the latest version from the official third party repository.

### CODE REVIEW

File: src/charm.py [hunk 1]
```diff
@@ -35,75 +34,101 @@ def _on_install(self, event):
...
-            check_call(["sudo", "apt-get", "update", "-qq"])
-
-            if os.path.exists("/var/lib/postgresql/12"):
-                pgver = 12
-            elif os.path.exists("/var/lib/postgresql/14"):
-                pgver = 14
+            debs = ["loader-deb", "tools-deb", "deb"]
+            deb_paths = []
+
+            for i in range(len(debs)):
```
Comment (frankban) [inline src/charm.py]: Just `for deb in debs`?
And then `path = self.model.resources.fetch(deb)` ?
like `paths = [self.model.resources.fetch(deb) for deb in debs]`

**Review by frankban — APPROVED** [review]
Looks good!

**Review by alesstimec — CHANGES_REQUESTED** [review]
I don't think this approach will work quite as expected.. 

The install event is only emitted once per unit at the beginning of a charm's lifecycle (https://juju.is/docs/sdk/install-event). If at that point the user has not yet specified debs as resource, charm will install packages from the default repo. If you provide debs at a later stage by using "juju attach-resource" the install hook will not fire again and your debs will not be installed.

Also a more standard approach is for a user to provide a PPA and have packages installed from it instead of the standard repo.

**Comment by TeodorPt** [comment]
> I don't think this approach will work quite as expected..
> 
> The install event is only emitted once per unit at the beginning of a charm's lifecycle (https://juju.is/docs/sdk/install-event). If at that point the user has not yet specified debs as resource, charm will install packages from the default repo. If you provide debs at a later stage by using "juju attach-resource" the install hook will not fire again and your debs will not be installed.
> 
> Also a more standard approach is for a user to provide a PPA and have packages installed from it instead of the standard repo.

Thanks for the input. Indeed, I did not consider adding resources post-install... I can add the custom PPA config for the charm, but I think debs could still be useful in case the charm is to be run in a very confined environment? Should we keep it around, with the caveat that it is a sub-par experience, or should we just remove it?

**Comment by alesstimec** [comment]
that is up to you.. you can keep the debs, but make it clear in the readme, that it may not always work - you need to attach resource before the install hook fires..

**Review by brotich — APPROVED** [review]

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -19,6 +19,8 @@ jobs:
         uses: actions/checkout@v3
       - name: Setup operator environment
         uses: charmed-kubernetes/actions-operator@main
+        with:
+          juju-channel: 3.1/stable
```
Comment (alesstimec) [inline .github/workflows/ci.yml:23]: oh, somebody likes to live on the edge ;)

**Comment by TeodorPt** [comment]
I've run a test with the new code changes in the following way:
- deployed the app configured to TimescaleDB version 2.9.3
- tested that TimescaleDB works on a database by enabling the extension, verified the version
- changed the config to use latest
- observed the packages are updated
- tested that TimescaleDB still works and also triggered an update for the extension, verified the version is correct

File: config.yaml [hunk 3]
```diff
@@ -1 +1,21 @@
...
+      The apt-repository to use for installing TimescaleDB.
+    type: string
+  apt-key:
+    default: https://packagecloud.io/timescale/timescaledb/gpgkey
+    description: |
+      The apt GPG key to use for installing TimescaleDB, in conjunction with
+      'apt-repository'. If it does not apply, configure to empty.
+    type: string
+  version:
+    default:
```
Comment (brotich) [inline config.yaml:14]: what does an empty default mean here?

**Review by brotich — APPROVED** [review]
looks good with my limited charm knowledge

**Review by alesstimec — COMMENTED** [review]
I would rather see the deb resources being handled in the update charm hook. Also instead of recording the "installed" flag in the state, i'd propose recording the hash of each of the deb package files provided. If the hash changes, you reinstall the package..

File: config.yaml (hunk 3 above)
Comment (TeodorPt) [reply config.yaml:14]: It means the latest version is to be installed.

File: config.yaml [hunk 4]
```diff
@@ -1 +1,17 @@
...
+  apt-repository:
+    default: https://packagecloud.io/timescale/timescaledb/ubuntu/
+    description: |
+      The apt-repository to use for installing TimescaleDB.
+    type: string
+  apt-key:
+    default: https://packagecloud.io/timescale/timescaledb/gpgkey
+    description: |
+      The apt GPG key to use for installing TimescaleDB, in conjunction with
+      'apt-repository'. If it does not apply, configure to empty.
```
Comment (alesstimec) [inline config.yaml:11]: "configure to empty" does not apply, as it has a meaningful default value.

**Review by alesstimec — APPROVED** [review]
LGTM!

File: config.yaml (hunk 4 above)
Comment (TeodorPt) [reply config.yaml:11]: By "configure to empty" I meant "set the config to an empty-string value".

**Pull request merged**